### PR TITLE
Implement /register/available API

### PR DESCRIPTION
**Pull request by MTRNord**
My first PR so if there is anything wrong with it please tell me ;) I hope I didn't miss anything.

Adds `GET /register/available?username=....`
I think it does what is expected by #284

### CODE REVIEW

File: src/github.com/matrix-org/dendrite/clientapi/routing/routing.go [hunk 1]
```diff
@@ -131,6 +131,10 @@ func Setup(
 		return writers.LegacyRegister(req, accountDB, deviceDB, &cfg)
 	})).Methods("POST", "OPTIONS")
 
+	r0mux.Handle("/register/available", common.MakeExternalAPI("registerAvailable", func(req *http.Request) util.JSONResponse {
+		return readers.RegisterAvailable(req, accountDB)
+	})).Methods("GET", "OPTIONS")
```
Comment (babolivier) [inline src/github.com/matrix-org/dendrite/clientapi/routing/routing.go]: Handling `OPTIONS` here isn't needed as CORS pre-flighted requests aren't sent for a `GET` request: https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS#Preflighted_requests

File: src/github.com/matrix-org/dendrite/clientapi/readers/register.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+			JSON: jsonerror.InvalidUsername("User ID can only contain characters a-z, 0-9, or '_-./'"),
+		}
+	} else if username[0] == '_' { // Regex checks its not a zero length string
+		return &util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.InvalidUsername("User ID can't start with a '_'"),
+		}
+	}
+	return nil
+}
```
Comment (erikjohnston) [inline src/github.com/matrix-org/dendrite/clientapi/readers/register.go]: This seems to be duplicated from `writers.register`? Either we should make he validation function in writers public, or we should move the `RegisterAvailable` func to `writers.register` (which is probably my preferred approach)

Comment (erikjohnston) [inline src/github.com/matrix-org/dendrite/clientapi/readers/register.go]: This seems to be duplicated from writers.register? Either we should make he validation function in writers public, or we should move the RegisterAvailable func to writers.register (which is probably my preferred approach)

Comment (MTRNord) [reply src/github.com/matrix-org/dendrite/clientapi/readers/register.go]: The function on writers.register is basicly te same except that one also checks the password. So either way I would atleast need to split the validate function in username and password

Comment (erikjohnston) [reply src/github.com/matrix-org/dendrite/clientapi/readers/register.go]: I think splitting up into `validateUserName` and `validatePassword` makes sense

**Pull request merged**